packages/catalog-model/src/entity/ref.ts

```diff
@@ -0,0 +1,146 @@
+ * otherwise specified as part of the options, and will throw an error if no
+ * kind was specified in the input reference and no default kind was given.
+ *
+ * @param ref The reference to parse
+ * @param context The context of defaults that the parsing happens within
+ * @returns A complete entity name
+ */
+export function parseEntityName(
+  ref: EntityRef,
+  context: EntityRefContext = {},
+): EntityName {
+  const { kind, namespace, name } = parseEntityRef(ref, {
+    defaultNamespace: 'default',
+    ...context,
+  });
+
+  if (!kind) {
+    throw new Error(
+      `Entity reference ${namespace}/${name} did not contain a kind`,
+    );
+  }
+
+  // This is a corner case - when the ref contained a namespace but it was the
+  // empty string (so the default did not kick in)
+  if (!namespace) {
+    throw new Error(
+      `Entity reference ${kind}:${name} did not contain a namespace`,
+    );
+  }
+
+  return {
+    kind,
+    namespace: namespace!,
+    name,
+  };
+}
+
+/**
+ * Parses an entity reference, either on string or compound form, and returns
+ * a structure with a name, and optional kind and namespace.
+ *
+ * The options object can contain default values for the kind and namespace,
+ * that will be used if the input reference did not specify any.
+ *
+ * @param ref The reference to parse
+ * @param context The context of defaults that the parsing happens within
+ * @returns The compound form of the reference
+ */
+export function parseEntityRef(
+  ref: EntityRef,
+  context: EntityRefContext = {},
+): CompoundEntityRef {
+  if (!ref) {
+    throw new Error(`Entity reference must not be empty`);
+  }
+
+  if (typeof ref === 'string') {
+    const match = /^(?:([^:/]+):)?(?:([^:/]+)\/)?([^:/]+)$/.exec(ref.trim());
+    if (!match) {
+      throw new Error(
+        `Entity reference "${ref}" was not on the form [<kind>:][<namespace>/]<name>`,
+      );
+    }
+
+    return {
+      kind: match[1] ?? context.defaultKind,
+      namespace: match[2] ?? context.defaultNamespace,
+      name: match[3],
+    };
+  }
+
+  const { kind, namespace, name } = ref;
+  if (kind === '') {
+    throw new Error('Entity reference kinds must not be empty');
+  } else if (namespace === '') {
+    throw new Error('Entity reference namespaces must not be empty');
+  } else if (!name) {
+    throw new Error('Entity references must contain a name');
+  }
+
+  return {
+    kind: kind ?? context.defaultKind,
+    namespace: namespace ?? context.defaultNamespace,
+    name,
+  };
+}
+
+/**
+ * Takes an entity reference or name, and outputs an entity reference on the
+ * most compact form possible. I.e. if the parts do not contain any
+ * special/reserved characters, it outputs the string form, otherwise it
+ * outputs the compound form.
+ *
+ * @param ref The reference to serialize
+ * @returns The same reference on either string or compound form
+ */
+export function serializeEntityRef(
```

Comment:
In case we don't allow those characters in the components, which afaik we don't? This could just be `stringifyEntityRef`, or `stringifyEntityHandle`.

Reply:
again, it's configurable, and i don't want to make assumptions throughout the entire codebase based on that